keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,364 @@
+* It provides the community a path to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. 
+* This feature would make it possible to create new types of CSI drivers such as ephemeral volume drivers.  They can be used to inject arbitrary states such as configuration, secrets, identity, variables or similar information inside pods. 
+* This can also provide a migration path for older Flex-style drivers (allowing the deprecation of Flex.)
+
+### Goals 
+* Provide a high level design for inline CSI volumes support
+* Define API changes needed to support this feature
+* Inline CSI volumes should work with existing CSI drivers
+* Design for ephemeral inline CSI volumes support
+* Ensure that inline CSI volumes usage is secure
+
+### Non-goals
+The followings will not be addressed by this KEP:
+* Introduce new CSI spec changes to support this feature
+* Introduce required changes to existing CSI drivers for this feature
+* Support for topology or pod placement scheme for ephemeral inline volumes
+* Support for PV/PVC related features such as topology, raw block, mount options, and resizing
+
+## User stories
+* As a storage provider, I want to be able to create CSI drivers that support persistent volumes that can be nested within pod specs.  These volumes would work similarly to how my current in-tree drivers work (with minor limitations addressed later).
+* As a storage provider, I want to use the CSI API to develop drivers that can mount ephemeral volumes that follow the lifecycles of pods where they are embedded.   This feature would allow me to create drivers that work similarly to how the in-tree Secrets or ConfigMaps driver works.  My ephemeral CSI driver should allow me to inject arbitrary data into a pod using a volume mount point. 
+* As a user, I want to be able to deploy pods, with persistent CSI volumes embedded in their specs, without the use of PV/PVC.  
+* As a user I want to be able to define pod specs with embedded ephemeral CSI volumes that are created/mounted when the pod is deployed and are deleted when the pod goes away.
+
+### Examples
+
+**Example 1**
+A pod spec with a persistent inline CSI volume.  Note that the `volumeHandle` is required and refers to a pre-existing volume.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: testpod
+spec:
+  containers:
+    ...
+  volumes:
+      - name: vol
+        csi:
+          driver: mock.storage.kubernetes.io
+          volumeAttributes:
+              name: "Mock Volume 1"
+          volumeHandle: "1"
+```
+
+**Example 2**
+A pod spec with an ephemeral inline CSI volume.  Note that because the volume is expected to be ephemeral, the `volumeHandle` is not provided.  Instead a CSI-generated ID will be submitted to the driver.
```

Comment:
what happens if the user specifies a volumeHandler in this case? is it plumbed through or overridden by a generated value? (in other words, can an ephemeral driver count on the volume handle being random or outside the pod-spec-author's control?)